### ISSUE TITLE
Change in the order between interceptors and mapped interceptors in Spring 5

### ISSUE THREAD

**Issue by mykevinjung** [spring-projects/spring-framework#25390]
When a handler mapping creates a `HandlerExecutionChain`, Spring 4.1.x added `adaptedInterceptors` first and then added `mappedInterceptors` as shown [here](https://github.com/spring-projects/spring-framework/blob/11fc10282d5fec21346bc541b564b1f45197f942/spring-webmvc/src/main/java/org/springframework/web/servlet/handler/AbstractHandlerMapping.java#L349-L356).

However, Spring 5.1.x merged adapted interceptors and mapped interceptors into `adaptedInterceptors` variable and `getHandlerExecutionChain` returns whatever in the order in `adaptedInterceptors` variable as shown [here](https://github.com/spring-projects/spring-framework/blob/14d539017cffbf534794dc9fbb8a9bc9b21ea07b/spring-webmvc/src/main/java/org/springframework/web/servlet/handler/AbstractHandlerMapping.java#L480-L490).  However, when the handler mapping populates `adaptedInterceptors`, it adds mapped interceptors into `adaptedInterceptors` first and then adds regular `interceptors` into `adaptedInterceptors` later as shown [here](https://github.com/spring-projects/spring-framework/blob/14d539017cffbf534794dc9fbb8a9bc9b21ea07b/spring-webmvc/src/main/java/org/springframework/web/servlet/handler/AbstractHandlerMapping.java#L296-L297).

As a result, mapped interceptors are executed first and then regular interceptors are executed later in Spring 5 while regular interceptors were executed first and then mapped interceptors were executed later in Spring 4.  Is this an intended change?  I noticed this while I am upgrading Spring version from 4.1.x to 5.1.x in my application.  How can I override in Spring 5 so that regular interceptors are executed first and then mapped interceptors executed after?

Maybe [AbstractHandlerMapping.initApplicationContext](https://github.com/spring-projects/spring-framework/blob/14d539017cffbf534794dc9fbb8a9bc9b21ea07b/spring-webmvc/src/main/java/org/springframework/web/servlet/handler/AbstractHandlerMapping.java#L293-L298) should be as following?

```
	@Override
	protected void initApplicationContext() throws BeansException {
		extendInterceptors(this.interceptors);
		initInterceptors(); // init and adapt this.interceptors first
		detectMappedInterceptors(this.adaptedInterceptors); // then adapt mapped interceptors
	}
```

**Comment by mykevinjung** [comment]
The following is a way that I can think to override the behavior.  But it's really hacking and I like to have any better way.  Any suggestion?  By the way, what was the reason of switching the order between interceptors and mapped interceptors?  Once I understand, maybe I can find some other way.

```
public class CustomRequestMappingHandlerMapping extends RequestMappingHandlerMapping {

    @Override
    protected void initApplicationContext() throws BeansException
    {
        // this will add interceptors to adaptedInterceptors first
        super.initInterceptors();

        // and then calling initApplicationContext,
        // then the overridden extendInterceptors will clear interceptors so that they aren't added twice in another initInterceptors() call
        // and detectMappedInterceptors will add mapped interceptors
        // and initInterceptors() won't do anything as interceptors is cleared
        super.initApplicationContext();
    }

    @Override
    protected void extendInterceptors(List<Object> interceptors) {
        interceptors.clear();
    }
}
```

**Comment by rstoyanchev** [comment]
This goes back to the changes for #17272 to allow more uniform treatment of pattern-mapped and regular interceptors in terms of ordering especially when configured through the MVC config.

**Comment by mykevinjung** [comment]
Thanks @rstoyanchev for your response.  I had a look https://github.com/spring-projects/spring-framework/issues/17272.  So the issue was that we wanted to put an order explicitly regardless of mapped interceptors or regular interceptors.  However, looks like the solution was switching the order between regular interceptors and mapped interceptors.  I am not convinced that the solution has really achieved the goal (put an order explicitly regardless of mapped interceptors or regular interceptors).  It might have solved the specific use case in the issue #17272 though.

**Comment by rstoyanchev** [comment]
Indeed, the solution for #17272 made it possible to maintain the order of interceptors provided through the Java config regardless of whether they are mapped or not. For interceptors provided directly to the `HandlerMapping`, it depends when they are provided relative to the Java config.

Can you register your interceptors through the Java config? Alternatively, the MVC Java config provides a way to supply the `RequestMappingHandlerMapping` instance by switching to [advanced config](https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html#mvc-config-advanced-java) and overriding a protected method. In Spring Boot you [can use](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-spring-mvc-auto-configuration) `WebMvcRegistrations` for the same.

**Comment by mykevinjung** [comment]
Thanks @rstoyanchev for the information. In our case, the app is kind of a legacy app and there are around 90 interceptors defined via xml definitions, so too many places to update.  I decided just to override `getHandlerExecutionChain()` as in the code [above](https://github.com/spring-projects/spring-framework/issues/25390#issuecomment-660672611).

If this works with Java config, then should be ok.  Closing.
